iroh-blobs/src/util/local_pool.rs

```diff
@@ -0,0 +1,698 @@
+            // will then immediately return.
+            self.send.send(Message::Finish).ok();
+        }
+        self.await_thread_completion().await;
+    }
+
+    fn threads_u32(&self) -> u32 {
+        self.threads
+            .len()
+            .try_into()
+            .expect("invalid number of threads")
+    }
+
+    async fn await_thread_completion(&self) {
+        // wait for all threads to finish.
+        // Each thread will add a permit to the semaphore.
+        let wait_for_semaphore = async move {
+            let _ = self
+                .shutdown_sem
+                .acquire_many(self.threads_u32())
+                .await
+                .expect("semaphore closed");
+        };
+        // race the semaphore wait with the cancel token in case somebody
+        // cancels the pool while we are waiting.
+        tokio::select! {
+            _ = wait_for_semaphore => {}
+            _ = self.cancel_token.cancelled() => {}
+        }
+    }
+}
+
+impl Drop for LocalPool {
+    fn drop(&mut self) {
+        self.cancel_token.cancel();
+        let current_thread_id = std::thread::current().id();
+        for handle in self.threads.drain(..) {
+            // we have no control over from where Drop is called, especially
+            // if the pool ends up in an Arc. So we need to check if we are
+            // dropping from within a pool thread and skip it in that case.
+            if handle.thread().id() == current_thread_id {
+                tracing::error!("Dropping LocalPool from within a pool thread.");
+                continue;
+            }
+            // Log any panics and resume them
+            if let Err(panic) = handle.join() {
+                let panic_info = get_panic_info(&panic);
+                let thread_name = get_thread_name();
+                tracing::error!("Error joining thread: {}\n{}", thread_name, panic_info);
+                std::panic::resume_unwind(panic);
+            }
+        }
+    }
+}
+
+/// Errors for spawn failures
+#[derive(thiserror::Error, Debug)]
+pub enum SpawnError {
+    /// Pool is shut down
+    #[error("pool is shut down")]
+    Shutdown,
+}
+
+type SpawnResult<T> = std::result::Result<T, SpawnError>;
+
+/// Future returned by [`LocalPoolHandle::run`] and [`LocalPoolHandle::try_run`].
+///
+/// Dropping this future will immediately cancel the task. The task can fail if
+/// the pool is shut down.
+#[repr(transparent)]
+#[derive(Debug)]
+pub struct Run<T>(tokio::sync::oneshot::Receiver<T>);
+
+impl<T> Run<T> {
+    /// Abort the task
+    ///
+    /// Dropping the future will also abort the task.
+    pub fn abort(&mut self) {
+        self.0.close();
+    }
+}
+
+impl<T> Future for Run<T> {
+    type Output = std::result::Result<T, SpawnError>;
+
+    fn poll(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        self.0.poll(cx).map_err(|_| SpawnError::Shutdown)
```

Comment:
should we log this error?

Reply:
It's not really an error. It is just what happens if you await a future that runs on a pool that has been shut down.

Reply:
I originally had this just return a oneshot receiver, but having my own wrapper allows implementing Into<io::Error> for it, which has some advantages (JoinError does the same).

Reply:
maybe add a comment, so it is clear for future readers